Source/WebCore/editing/markup.cpp

```diff
@@ -1481,16 +1478,14 @@ ExceptionOr<void> replaceChildrenWithFragment(ContainerNode& container, Ref<Docu
     }
 
     // We don't Use RefPtr here because canUseSetDataOptimization() below relies on the
-    // containerChild's ref count.
-    WeakPtr containerChild = containerNode->firstChild();
-    if (containerChild && !containerChild->nextSibling()) {
-        if (is<Text>(*containerChild) && hasOneTextChild(fragment) && canUseSetDataOptimization(downcast<Text>(*containerChild), mutation)) {
-            Ref { downcast<Text>(*containerChild) }->setData(downcast<Text>(*fragment->firstChild()).data());
+    // containerChildText's ref count.
+    auto* containerChildText = dynamicDowncast<Text>(containerNode->firstChild());
+    if (containerChildText && !containerChildText->nextSibling()) {
+        if (RefPtr fragmentChild = dynamicOnlyTextChild(fragment); fragmentChild && canUseSetDataOptimization(*containerChildText, mutation)) {
+            Ref { *containerChildText }->setData(fragmentChild->data());
             return { };
         }
-
-        Ref protectedContainerChild { *containerChild };
-        return containerNode->replaceChild(fragment, protectedContainerChild);
+        return containerNode->replaceChild(fragment, Ref { *containerChildText });
```

Comment:
This was quite tricky to get correct and ensure existing tests kept passing. It needs careful review.